### DISCUSSION TITLE
Supporting multiple LLMs in Spring Boot auto configuration

### DISCUSSION THREAD

**Discussion by ddobrin** [langchain4j/langchain4j#2257] [Ideas]
Hello everybody,
bringing up the following idea as a discussion, to coalesce around a Spring Boot auto configuration change in `langchain4j-spring`. Discussions give visibility to larger scoped changes.

**Background**
AI platforms such as Vertex AI and others support the configuration of multiple models in a model-as-a-service configuration. Different private or open-models can be configured and accessed in the same AI platform.

A scenario, for example:
```
...
langchain4j.google-ai-gemini.chatModel.model-name=gemini-1.5-flash
langchain4j.google-ai-gemini.chatModel.temperature=0.5
langchain4j.google-ai-gemini.chatModel.topP=0.5
...
langchain4j.google-ai-gemini.chatModel.model-name=claude-3-5-sonnet-20240620
langchain4j.google-ai-gemini.chatModel.temperature=0.9
langchain4j.google-ai-gemini.chatModel.topP=0.9
...
```
can't be supported via auto-configuration at this time.

**Proposal**
I am proposing a change in the configuration to support multiple models with the naming:

```
...
langchain4j.google-ai-gemini.<model-family-name>.chatModel.model-name=
langchain4j.google-ai-gemini.<model-family-name>.chatModel.temperature=
langchain4j.google-ai-gemini.<model-family-name>.chatModel.topP=
...
```

For example:
```
...
langchain4j.google-ai-gemini.gemini-flash.gemini.chatModel.model-name=gemini-1.5-flash
langchain4j.google-ai-gemini.gemini-flashgemini.chatModel.temperature=0.5
langchain4j.google-ai-gemini.gemini.flash-chatModel.topP=0.5
...
langchain4j.google-ai-gemini.gemini-pro.gemini.chatModel.model-name=gemini-1.5-pro
langchain4j.google-ai-gemini.gemini-pro.gemini.chatModel.temperature=0.5
langchain4j.google-ai-gemini.gemini.pro.chatModel.topP=0.5
...
langchain4j.google-ai-gemini.sonnet.chatModel.model-name=claude-3-5-sonnet-20240620
langchain4j.google-ai-gemini.sonnet.chatModel.temperature=0.9
langchain4j.google-ai-gemini.sonnet.chatModel.topP=0.9
...
```

**Comment by Unknown-Pika** [comment]
How would you dynamically create or write a configuration that creates N number of beans?  
The idea looks good though
Also i am curious of the use cases of the same

**Comment by ddobrin** [comment]
You can use a Spring Boot starter for configuration or the equivalent for Quarkus and @Configure respectively @Inject a qualifier for "model-family-name"

Some use cases, for example, are:

1. a model is used for faster operations, while a more powerful, yet slower model, is used for the final LLM operation (ex.: [hypothetical questions embedding](https://github.com/datastaxdevs/conference-2024-devoxx/blob/main/devoxx-rag-naive-to-advanced/src/test/java/devoxx/rag/_3_advanced_rag_ingestion/_37_hypothetical_questions_embedding.java) from @glaforge 
2. evaluation: you get a summary from a model and validate the summary against a different model, providing summary and original text to the second model
3. code generation improvements: similar to the previous however for code generation

There are more use-cases.
